src/core/bonded_interactions/angle_cosine.hpp

```diff
@@ -46,9 +46,10 @@ int angle_cosine_set_params(int bond_type, double bend, double phi0);
  *  @return Forces on the second, first and third particles, in that order.
  */
 inline std::tuple<Utils::Vector3d, Utils::Vector3d, Utils::Vector3d>
-calc_angle_cosine_3body_forces(Particle const *p_mid, Particle const *p_left,
-                               Particle const *p_right,
-                               Bonded_ia_parameters const *iaparams) {
+calc_angle_cosine_3body_forces(Particle const *const p_mid,
```

Comment:
actually I think we should not have particles as arguments but only their positions.

Reply:
this means exposing the internal implementation of particles in `pressure_inline.hpp`:
```diff
   switch (iaparams->type) {
   case BONDED_IA_ANGLE_HARMONIC:
     std::tie(f_mid, f_left, f_right) =
-        calc_angle_harmonic_3body_forces(p_mid, p_left, p_right, iaparams);
+        calc_angle_harmonic_3body_forces(p_mid->r.p, p_left->r.p, p_right->r.p, iaparams);
     break;
```

Reply:
i think all these functions don't actually need a particle but really just a position

Reply:
There is issue #3034 for this, can be dealt with later.